File: 721-accounts-merge/accounts-merge.py
class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        adj = collections.defaultdict(list)
        for i, a in enumerate(accounts):
            for email in a[1:]:
                adj[email].append(i)
        visited = set()
        def dfs(idx, emails):
            if idx in visited: return
            visited.add(idx)
            for email in accounts[idx][1:]:
                emails.add(email)
                for nei in adj[email]:
                    dfs(nei, emails)

        res = []
        for i, a in enumerate(accounts):
            if i in visited: continue
            name, emails = a[0], set()
            dfs(i, emails)
            res.append([name] + sorted(emails))
        return res
            
"""
dfs
johnsmith@mail.com: [0]
john_newyork@mail.com: [0]
john00@mail.com: [0, 1]
mary@mail.com: [2]
johnnybravo@mail.com: [3]
"""

File: 721-accounts-merge/accounts-merge.js
/**
 * @param {string[][]} accounts
 * @return {string[][]}
 */
class UnionFind {
    constructor(N) {
        this.parents = Array(N).fill(1).map((e, idx) => idx)
        this.ranks = Array(N).fill(1)
    }
    
    find(n) {
        if (this.parents[n] === n) return n
        this.parents[n] = this.find(this.parents[this.parents[n]])
        return this.parents[n]
    }

    union(n1, n2) {
        const [p1, p2] = [this.find(n1), this.find(n2)]
        if (p1 === p2) return false
        if (this.ranks[p1] >= this.ranks[p2]) {
            this.ranks[p1] += this.ranks[p2]
            this.parents[p2] = p1
        } else {
            this.ranks[p2] += this.ranks[p1]
            this.parents[p1] = p2
        }
        return true
    }
}
var accountsMerge = function(accounts) {
    const emailToIdx = {} // email: accountIdx
    const uf = new UnionFind(accounts.length)
    accounts.forEach((account, idx) => {
        const emails = account.slice(1)
        emails.forEach(email => {
            if (email in emailToIdx) {
                uf.union(idx, emailToIdx[email])
            }
            emailToIdx[email] = uf.find(idx)
        })
    })

    const parentIdxToIndices = {} // parentIdx: [idx1, idx2]
    accounts.forEach((account, idx) => {
        const parentIdx = uf.find(idx)
        if (!(parentIdx in parentIdxToIndices)) {
            parentIdxToIndices[parentIdx] = new Set()
        }
        parentIdxToIndices[parentIdx].add(idx)
    })
    
    const res = []
    for (const [parentIdx, indices] of Object.entries(parentIdxToIndices)) {
        const account = accounts[Number(parentIdx)]
        const name = account[0]
        const emails = []
        for (const idx of indices) {
            const childAccount = accounts[idx].slice(1)
            emails.push(...childAccount)
        }
        const uniqueEmails = new Set(emails)
        const sortEmailAccounts = [...uniqueEmails]
        sortEmailAccounts.sort()
        res.push([name, ...sortEmailAccounts])
    }
    return res
};